sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ProjectFilterInAggregates.scala

```diff
@@ -0,0 +1,193 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeReference, Expression, If, IsNotNull, Literal, NamedExpression}
+import org.apache.spark.sql.catalyst.expressions.aggregate.{AggregateExpression, AggregateFunction}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * If an aggregate query with filter clause, this rule will create a project node so as to filter
+ * the output of aggregate's child in advance.
+ *
+ * First example: query with filter clauses (in sql):
+ * {{{
+ *   val data = Seq(
+ *     (1, "a", "ca1", "cb1", 10),
+ *     (2, "a", "ca1", "cb2", 5),
+ *     (3, "b", "ca1", "cb1", 13))
+ *     .toDF("id", "key", "cat1", "cat2", "value")
+ *   data.createOrReplaceTempView("data")
+ *
+ *   SELECT
+ *     COUNT(DISTINCT cat1) AS cat1_cnt,
+ *     COUNT(DISTINCT cat2) FILTER (WHERE id > 1) AS cat2_cnt,
+ *     SUM(value) AS total,
+  *    SUM(value) FILTER (WHERE key = "a") AS total2
+ *  FROM
+ *    data
+ *  GROUP BY
+ *    key
+ * }}}
+ *
+ * This translates to the following (pseudo) logical plan:
+ * {{{
+ * Aggregate(
+ *    key = ['key]
+ *    functions = [COUNT(DISTINCT 'cat1),
+  *                COUNT(DISTINCT 'cat2) with FILTER('id > 1),
+ *                 SUM('value),
+  *                SUM('value) with FILTER('key = "a")]
+ *    output = ['key, 'cat1_cnt, 'cat2_cnt, 'total, 'total2])
+ *   LocalTableScan [...]
+ * }}}
+ *
+ * This rule rewrites this logical plan to the following (pseudo) logical plan:
+ * {{{
+ *   Aggregate(
+ *      key = ['key]
+ *      functions = [count('_gen_attr_1),
+  *                  count('_gen_attr_2) with FILTER('_gen_attr_2 is not null),
+ *                   sum('_gen_attr_3),
+  *                  sum('_gen_attr_4) with FILTER('_gen_attr_4 is not null)]
+ *      output = ['key, 'cat1_cnt, 'cat2_cnt, 'total, 'total2])
+ *     Project(
+ *        projectionList = ['key,
+ *                          'cat1,
+  *                         if ('id > 1) 'cat2 else null,
+ *                          cast('value as bigint),
+ *                          if ('key = "a") cast('value as bigint) else null]
+ *        output = ['key, '_gen_attr_1, '_gen_attr_2, '_gen_attr_3, '_gen_attr_4])
```

Comment:
`cat1` is not related to the filter, why do we change its name to `_gen_attr_1`?

Reply:
For convenience and unification, we always alias the column, even if there is no filter.